core/domain/image_services_test.py

```diff
@@ -0,0 +1,128 @@
+
+    TEST_IMAGE_WIDTH = 3000
+    TEST_IMAGE_HEIGHT = 2092
+
+    def setUp(self):
+        super(ImageServicesUnitTests, self).setUp()
+        with python_utils.open_file(
+            os.path.join(feconf.TESTS_DATA_DIR, 'dummyLargeImage.jpg'),
+            mode='rb', encoding=None) as f:
+            self.jpeg_raw_image = f.read()
+        with python_utils.open_file(
+            os.path.join(feconf.TESTS_DATA_DIR, 'img.png'),
+            mode='rb', encoding=None) as f:
+            self.png_raw_image = f.read()
+
+    def test_image_dimensions_outputs_correctly(self):
+        height, width = (
+            image_services.get_image_dimensions(self.jpeg_raw_image))
+        self.assertEqual(self.TEST_IMAGE_HEIGHT, height)
+        self.assertEqual(self.TEST_IMAGE_WIDTH, width)
+
+    def test_dev_mode_returns_regular_image(self):
+        compressed_image = (image_services.compress_image(
+            self.jpeg_raw_image, 0.8))
+        height, width = (
+            image_services.get_image_dimensions(compressed_image))
+        self.assertEqual(self.TEST_IMAGE_HEIGHT, height)
+        self.assertEqual(self.TEST_IMAGE_WIDTH, width)
+
+    def test_compress_image_returns_correct_dimensions(self):
+        with self.swap(constants, 'DEV_MODE', False):
+            compressed_image = (
+                image_services.compress_image(self.jpeg_raw_image, 0.5))
+        height, width = (
+            image_services.get_image_dimensions(compressed_image))
+        self.assertEqual(self.TEST_IMAGE_HEIGHT * 0.5, height)
+        self.assertEqual(self.TEST_IMAGE_WIDTH * 0.5, width)
+
+    def test_invalid_scaling_factor_triggers_value_error(self):
+        email_exception = self.assertRaisesRegexp(
+            ValueError, 'Scaling factor should be less than 1.')
+        with self.swap(constants, 'DEV_MODE', False), email_exception:
+            image_services.compress_image(self.jpeg_raw_image, 1.1)
+
+    def test_compression_results_in_correct_format(self):
+        with self.swap(constants, 'DEV_MODE', False):
+            compressed_image = (
+                image_services.compress_image(self.jpeg_raw_image, 0.7))
+        pil_image = Image.open(io.BytesIO(compressed_image))
+        self.assertEqual(pil_image.format, 'JPEG')
+        with self.swap(constants, 'DEV_MODE', False):
+            compressed_image = (
+                image_services.compress_image(self.png_raw_image, 0.7))
+        pil_image = Image.open(io.BytesIO(compressed_image))
+        self.assertEqual(pil_image.format, 'PNG')
+
+    def test_compression_results_in_identical_files(self):
+        def is_different(file1, file2):
+            """Checks the differences between the two images and returns if they
+            are different.
+
+            Args:
+                file1: str. Content of first image.
+                file2: str. Content of second image.
+
+            Returns:
+                bool. Returns whether the images are different.
```

Comment:
bool. Whether the images are different.

Reply:
Not done (the point of what I wrote was to not include the "Returns").

Also see above -- this looks wrong?

Reply:
What the heck, I remember distinctly that I changed it... I must've accidentally checked out that change when I was messing with git.